File: src/test/java/org/hisp/dhis/lib/expression/util/RuleVariableValue.java
package org.hisp.dhis.lib.expression.util;

import lombok.Builder;
import lombok.Value;
import org.hisp.dhis.lib.expression.spi.ValueType;
import org.hisp.dhis.lib.expression.spi.VariableValue;

import java.util.List;

@Value
@Builder(toBuilder = true, builderMethodName = "of")
public class RuleVariableValue implements VariableValue {

    ValueType valueType;
    String value;
    List<String> candidates;
    String eventDate;

    @Override
    public String value() {
        return value;
    }

    @Override
    public Object valueOrDefault() {
        switch (valueType) {
            case NUMBER: return 0d;
            case DATE: return "2010-01-01";
            case BOOLEAN: return false;
            default: return "";
        }
    }

    @Override
    public List<String> candidates() {
        return candidates;
    }

    @Override
    public String eventDate() {
        return eventDate;
    }

    @Override
    public ValueType valueType() {
        return valueType;
    }
}

File: src/jvmTest/java/org/hisp/dhis/lib/expression/util/RuleVariableValue.kt
package org.hisp.dhis.lib.expression.util

import org.hisp.dhis.lib.expression.spi.ValueType
import org.hisp.dhis.lib.expression.spi.VariableValue

data class RuleVariableValue (
    private val valueType: ValueType,
    private val value: String?,
    private val candidates: List<String>?,
    private val eventDate: String?
) : VariableValue {

    constructor(valueType: ValueType) : this(valueType, null, null, null)

    override fun value(): String? {
        return value
    }

    override fun valueOrDefault(): Any {
        return when (valueType) {
            ValueType.NUMBER -> 0.0
            ValueType.DATE -> "2010-01-01"
            ValueType.BOOLEAN -> false
            else -> ""
        }
    }

    override fun candidates(): List<String>? {
        return candidates
    }

    override fun eventDate(): String? {
        return eventDate
    }

    override fun valueType(): ValueType {
        return valueType
    }
}
